### PR TITLE
Assignment1a

### PR DESCRIPTION
**Pull request by ischo0812**

### CODE REVIEW

**Comment by SwatiSuri** [comment]
He has created all the classes as requested. ..worked on his code to ensure no errors existed and all his tests run perfectly.

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (mab105120) [inline .vscode/launch.json:1]: Usually files that are specific to a developer setup (e.g., anything under .vscode) should not be committed/pushed to a remote repository. Since other developers who clone this code might have their own local settings which this would override. It is best practice to keep these files from push.

File: app/src/q_Gradecalc.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+# Create a function called: calculate_avg_grade
```
Comment (mab105120) [inline app/src/q_Gradecalc.py:1]: Any reason why your renamed the files to start with q_ ? How you name objects (i.e., modules, functions, classes, variables, etc.) is very important for code readability and maintainability. Properly named code does not need any documentation and can read smoothly. Avoid using vague naming conventions to not confuse others that are reading your code.

File: app/src/q_Gradecalc.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+# Returns: the function should return a dictionary with student ID as key and student average grade as value
+#
+# Example:
+# input: [Student('s1',20,1824,90.5), Student('s2',21,1823,80.0), Student('s1',20,1824,70.0)]
+# output: { 1824: 80.25, 1823: 80.0 }
+#
+# If the list of students is empty, return an empty dictionary
+# Make sure that you add type hints to the function paramter and return value
+
+from app.src.q_Student import Student as StatSheet
```
Comment (mab105120) [inline app/src/q_Gradecalc.py:13]: Why did you rename the Student class as StatSheet. I see that you're using StateSheet across the entire file, which is confusing. The function should expect a list of Student objects and calculate their average grade, if you use StatSheet as the type hint it will confuse the reader and make your code vague and hard to understand.

File: app/src/q_Gradecalc.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    if len(StatSheet) == 0:
+        raise Exception('Dictionary is empty')
+    else:
+        for statsheet in statsheets:
+            student_id: int = statsheet.student.id
+            if student_id in student_to_stats.keys():
+                student_to_stats.get(pstudent_id).append(statsheet)
+            else:
+                student_to_stats[student_id] = [statsheet]
+                student_to_avgpts: dict[int, float] = {}
```
Comment (mab105120) [inline app/src/q_Gradecalc.py:28]: This line needs to be indented out to remove it from under the else code block. Otherwise the variable scope would be limited to the else block and you the interpreter won't have access to this variable after it finishes executing the else block.

File: app/src/q_Person.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    # Class has two attributes 1.name(srt) 2.age(int)
+    def __init__(self, name: str, age: int) -> None:
+        self.name = name
+        self.age = age
+
+    def print_name(self) -> None:
+        print(f'{self.name} {self.age}')
+
+
+# p1 = Person(name='Abraham', age=31)
```
Comment (mab105120) [inline app/src/q_Person.py:16]: Please clean out unnecessary comments before pushing files to a remote repository. It is very easy for your code to get buried with stale and unnecessary comments, which makes it harder to read.

File: app/src/q_StatSheet.py [hunk 6]
```diff
@@ -0,0 +1,8 @@
+from app.src.q_Student import Student
```
Comment (mab105120) [inline app/src/q_StatSheet.py:1]: This class is not required. I think you are using the classroom example. The sample code from the class, while similar in functionality, is not the same as the assignment. Avoid following closely to reduce confusion.

File: app/src/q_Stats.py [hunk 7]
```diff
@@ -0,0 +1,5 @@
+class Stats():
```
Comment (mab105120) [inline app/src/q_Stats.py:1]: This class is not needed for this assignment.

File: app/src/q_Student.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+class Student(Person):
+    # Class has two attributes 1.id(int) 2.grade(float)
+    def __init__(self, name: str, age: int, id: int, grade: float) -> None:
+        Person.__init__(self, name, age)
+        self.id = id
+        self.grade = grade
+
+    def print_name(self) -> None:
+        print("ID : {2}, Name : {0}, Age : {1}, Grade : {3}".format(
```
Comment (mab105120) [inline app/src/q_Student.py:20]: Use the f'{<var_name>}' syntax for formatting text. It is much cleaner and easier to read.

File: app/src/q_Utils.py [hunk 9]
```diff
@@ -3,7 +3,27 @@
...
+
+def calculate_avg(numbers: list[t.Union[int, float]]) -> float:
+
+    sum: float = 0.0
+    if len(numbers) == 0:
+        raise Exception('expected a non-empty list')
+    else:
+        for grade in numbers:
+            sum += grade
+    return sum / len(numbers)
```
Comment (mab105120) [inline app/src/q_Utils.py:22]: Missing requirement to round the answer to the closest second decimal.

File: app/src/q_Utils.py [hunk 10]
```diff
@@ -3,7 +3,27 @@
...
+
+
+# numbers = [7.35, 9.9909, 105.33987]
```
Comment (mab105120) [inline app/src/q_Utils.py:25]: Please clean up un-used comments.

File: app/tests/Game.py [hunk 11]
```diff
@@ -0,0 +1,3 @@
+class game():
```
Comment (mab105120) [inline app/tests/Game.py:1]: Class not necessary for the assignment.

File: app/tests/Player.py [hunk 12]
```diff
@@ -0,0 +1,6 @@
+class Player():
```
Comment (mab105120) [inline app/tests/Player.py:1]: Class not necessary for the assignment.

File: app/tests/StatSheet.py [hunk 13]
```diff
@@ -0,0 +1,8 @@
+from app.src.Player import Player
```
Comment (mab105120) [inline app/tests/StatSheet.py:1]: class not required for assignment.

File: app/tests/Stats.py [hunk 14]
```diff
@@ -0,0 +1,5 @@
+class Stats():
```
Comment (mab105120) [inline app/tests/Stats.py:1]: class not required for assignment.

File: app/tests/Utils.py [hunk 15]
```diff
@@ -0,0 +1,13 @@
+
```
Comment (mab105120) [inline app/tests/Utils.py:1]: Not sure why you have two files that do the same thing.
Please use lower case to name modules: utils not Util.py
You only use upper case when naming a file that will define a class.
